File: projects/algorithms-and-leetcode/leetcode/bit_manipulation/python/power_of_two.py
"""
Given an integer n, return  true if it is a power of two. Otherwise, return false. An integer n is a power of two, if there exists an integer x such  that n == 2^x.
"""

from typing import List, Dict

class Solution:
    def isPowerOfTwo(self, n:int) -> bool:
        if n <= 0:
            return False
        if n & (n -1) == 0: # this checks if n is a power of two
            return True
        return False
    


if __name__ == "__main__":
    sol = Solution()
    print(sol.isPowerOfTwo(3))   # True

    
    

File: projects/algorithms-and-leetcode/leetcode/bit_manipulation/c++/power_of_two.cpp
// Given an integer n, return  true if it is a power of two. Otherwise, return false. An integer n is a power of two, if there exists an integer x such  that n == 2^x.

#include <iostream>
using namespace std;

class Solution{
    public:
    bool isPowerOfTwo(int n) {
        if (n <= 0) return false; // Handle non-positive integers
        return (n & (n -1)) == 0; // bit manipulation , if n is a power of two, it has only one bit set in its binary representation
    }
};


int main() {
    Solution sol;
    // Test cases
    std::cout << sol.isPowerOfTwo(1) << std::endl;   // True
    std::cout << sol.isPowerOfTwo(16) << std::endl;  // True
    std::cout << sol.isPowerOfTwo(18) << std::endl;  // False
    std::cout << sol.isPowerOfTwo(0) << std::endl;   // False
    std::cout << sol.isPowerOfTwo(-2) << std::endl;  // False
}